bootstrap/src/basis/file.ml

```diff
@@ -0,0 +1,243 @@
+open Rudiments
+open Result
+
+type t = {
+  mutable fd: Unix.file_descr option;
+}
+type e = string
+
+let buflen = 1024
+
+module Path = String
+
+module Mode = struct
+  type t =
+  | R_O
+  | W
+  | W_A
+  | W_AO
+  | W_C
+  | W_O
+  | RW
+  | RW_A
+  | RW_AO
+  | RW_C
+  | RW_O
+
+  let to_unix_flags t =
+    let open Unix in
+    match t with
+    | R_O   -> [O_RDONLY]
+    | W     -> [O_WRONLY; O_CREAT]
+    | W_A   -> [O_WRONLY; O_APPEND; O_CREAT]
+    | W_AO  -> [O_WRONLY; O_APPEND]
+    | W_C   -> [O_WRONLY; O_CREAT; O_EXCL]
+    | W_O   -> [O_WRONLY]
+    | RW    -> [O_RDWR; O_CREAT]
+    | RW_A  -> [O_RDWR; O_APPEND; O_CREAT]
+    | RW_AO -> [O_RDWR; O_APPEND]
+    | RW_C  -> [O_RDWR; O_CREAT; O_EXCL]
+    | RW_O  -> [O_RDWR]
+end
+
+let of_path ?(mode=Mode.RW) path =
+  let open Unix in
+  let unix_flags = Mode.to_unix_flags mode in
+  try
+    Ok ({fd=Some (openfile path unix_flags 0o640)})
+  with
+  | Unix_error (errno, fn_name, fn_param) -> begin
+    Error (String.concat ~sep:" -> " [Unix.error_message errno; fn_name; fn_param])
+  end
+  | _ -> not_reached ()
```

Comment:
So, this is a good example of where I'd like to improve things. The contents of this function look nothing like what Hemlock will look like. I'd like to hide the `Unix.openfile` call and exception handler behind a wrapper that looks more Hemlock-like, but not really sure what the wrapper function signature should look like. I also wonder if it'd be easier to wrap native c function calls than to wrap `Unix` calls.

Reply:
Writing native wrappers around the system calls sounds like a pretty good idea, actually.  The result would be quite close to how the Hemlock code will look.